File: src/leetcode/leetcode_109.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedListToBST(self, head: ListNode) -> TreeNode:

        def solve(root: TreeNode, s: ListNode, e: ListNode, flag: int):
            if root==None or s==e or s==None:
                return

            size = 0
            p = s
            while p != e:
                size += 1
                p = p.next
            p = s
            index = 0
            while index < size//2:
                index += 1
                p = p.next

            if flag == -1:
                root.left = TreeNode(p.val)
                solve(root.left, s, p, -1)
                solve(root.left, p.next, e, 1)

            else:
                root.right = TreeNode(p.val)
                solve(root.right, s, p, -1)
                solve(root.right, p.next, e, 1)



        if head == None:
            return None

        size = 0
        p = head
        while p != None:
            size += 1
            p = p.next

        p = head
        index = 0
        while index < size//2:
            index += 1
            p = p.next

        root = TreeNode(p.val)
        solve(root, head, p, -1)
        solve(root, p.next, None, 1)

        return root

File: src/leetcode/leetcode_109.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        if(head == NULL) return NULL;
        //处理第一个[head, null)
        // 先zh
        int size = 0;
        ListNode *p = head;

        while(p != NULL){
            size++;
            p = p->next;
        }
        if(size == 1) return new TreeNode(head->val);

        int index = 0;
        p = head;
        while(index<size/2 && p!=NULL){
            index++;
            p = p->next;
        }
        TreeNode *root = new TreeNode(p->val);
        solve(root, head, p, -1);
        solve(root, p->next, NULL, 1);

        return root;

    }
    void solve(TreeNode *root, ListNode *s, ListNode *e, int flag){
        if(root==NULL || s==e || s==NULL) return;

        int size = 0;
        ListNode *p = s;
        while(p != e){
            size++;
            p = p->next;
        }
        p = s;
        int index = 0;
        while(index<size/2 && p!=NULL){
            index++;
            p = p->next;
        }
        switch(flag){
            case -1:
                root->left = new TreeNode(p->val);
                solve(root->left, s, p, -1);
                solve(root->left, p->next, e, 1);
                break;
            default:
                root->right = new TreeNode(p->val);
                solve(root->right, s, p, -1);
                solve(root->right, p->next, e, 1);
        }
    }
};